packages/adapter/addon/-private/build-url-mixin.ts

```diff
@@ -12,6 +11,113 @@ type SnapshotRecordArray = import('@ember-data/store/-private/system/snapshot-re
   @module @ember-data/adapter
 */
 
+/*
+ The structure of this file is such because typing Mixins is hard. Here we've structured it in
+ such a way as to maximize the type information that a consumer can utilize. There are simpler
+ ways to type a mixin but we would not be able to provide the nice overload signature for buildURL
+*/
+// the interface must fully declare the function signatures that the individual functions
+// will also declare. If instead we try to keep them in sync by doing something like
+// `interface BuildURLMixin { buildURL: typeof buildURL }`
+// then an extending class overwriting one of the methods will break because typescript
+// thinks it is a switch from an instance prop (that is a method) to an instance method.
+interface BuildURLMixin {
```

Comment:
Would love to get @rwjblue or @chriskrycho to take a look here

Reply:
I talked this over with Chris but yes, would still like his review.

Reply:
Yeah, it's messy but until we're able to move away from mixins entirely (…could be a while 😬) I think this is least sucky option. It gives us the benefit of generally making sure we're keeping things in sync at the type level, providing decent user feedback via the overrides, and limiting where the damage of the casts has to be done. Mixins have *always* been the part of the system least amenable to type safety, precisely because it's very difficult to teach TypeScript what the type will be—which is fair, because the type in general can be arbitrarily dynamic.